### PR TITLE
fix(mimo-v2): reduce audio encoder attention over the attn-TP group

### LINKED ISSUE #37059: [Bug] MiMo-V2.5: any request with audio hangs the whole server under --enable-dp-attention (audio encoder reduces over the wrong TP group)
### Describe the bug

Serving **MiMo-V2.5** (`MiMoV2ForCausalLM`) with DP attention — the configuration the model effectively mandates on an 8-GPU node, since its fused `qkv_proj` requires effective attention TP = 4:

```
--tp 8 --dp 2 --enable-dp-attention --enable-dp-lm-head --mm-enable-dp-encoder --quantization fp8
```

any request **containing audio** permanently hangs the whole server. Text-only and video-without-audio requests work fine.

| case | result |
|---|---|
| text only | ✅ |
| video without audio | ✅ |
| **audio (any request carrying an audio item)** | ❌ hangs 600 s, then the server dies |
| everything after that | ❌ connection refused |

Server log at the moment of death:

```
[PG ID 4 PG GUID 23(attention_tp:device) Rank 3] Process group watchdog thread terminated with exception:
[Rank 3] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=14708, OpType=_ALLGATHER_BASE,
NumelIn=38144, NumelOut=152576, Timeout(ms)=600000) ran for 600083 milliseconds before timing out.
```

followed by gloo `Connection closed by peer` on the sibling ranks. Reproduced twice independently (official FP8 weights and a bf16 export) on 8×H200; identical signature both times.

### Root cause

`AudioEncoderAttention` builds its `VisionAttention` without `use_data_parallel` / `use_dp_attention_reduce` (`python/sglang/srt/models/mimo_audio.py:503`). Consequently:

- The audio encoder's attention weights **are sharded over the attention-TP group** — `_remap_audio_tokenizer_state_dict` slices by `attn_tp_rank`/`attn_tp_size` (`mimo_audio.py:1122-1173`), and `VisionAttention` defaults to `tp_size = attn_tp_size` (`layers/attention/vision.py:1072`).
- But its output projection's `RowParallelLinear` reduces over the **full TP group**: with `use_dp_attention_reduce=False` the reduce path takes `tensor_model_parallel_all_reduce` (`layers/linear.py:1666`) — 8 ranks, not 4.
- The audio encoder is only entered by ranks whose local batch actually has audio items: `general_mm_embed_routine` skips modalities with zero items (`managers/mm_utils.py:420`, `if len(items) != 0:`).

So when one DP group's batch carries audio and the other's does not, the audio-bearing group issues `encoder_layers × all_reduce` on the full-TP communicator that the other group never joins. The TP communicator desynchronizes and every subsequent collective piles up behind it.

The watchdog blames `attention_tp` only because that is the sole c10d-instrumented communicator in this setup: the full TP group runs through custom-allreduce/pynccl (no watchdog), while `attention_tp` falls through to `torch.distributed.all_gather_into_tensor` (`distributed/parallel_state.py:2596-2608`, `1246-1248`). The reported `_ALLGATHER_BASE` (`NumelIn=38144, NumelOut=152576` = 1 row × vocab 152576 / attn_tp 4) is the last-token logits gather of the audio request's prefill (`layers/logits_processor.py:926`) — enqueued behind the wedged all-reduce, never executed.

Two further consequences worth noting:

1. **Even when every DP group happens to have audio items, the numerics are wrong**: the full-TP reduce sums partials from unrelated DP groups' audio batches.
2. Video is immune precisely because the ViT honors `--mm-enable-dp-encoder` (`models/mimo_vl.py:261` → `use_data_parallel=True`, no collectives); the audio encoder is not wired to either flag.

Every other multimodal encoder in the tree already passes one of the two flags (`qwen3_vl.py:161`, `kimi_k25.py:130`, `minimax_vl_common.py:108`, `mimo_vl.py:300`); the audio encoder is the one that was missed. `mimo_v2_asr.py` inherits the same defect through `AudioEncoderMixin`.

Note `sglang`'s own `--watchdog-timeout` does not help — the 600 s timeout is `ProcessGroupNCCL`'s collective timeout, not the scheduler watchdog.

### Reproduction

```bash
python -m sglang.launch_server --model-path XiaomiMiMo/MiMo-V2.5 --trust-remote-code \
  --tp 8 --dp 2 --enable-dp-attention --enable-dp-lm-head --mm-enable-dp-encoder \
  --quantization fp8 --attention-backend fa3 --mm-attention-backend fa3
# then send any /v1/chat/completions request whose content includes an audio item
# (audio_only, or video with an audio track). With dp 2, the request lands in one
# DP group and the hang is immediate and deterministic in our runs.
```

### Environment

8×H200 (141 GB), sglang 0.5.18 (also verified the defective code path is unchanged on current `main` @ 97781eb7f), torch 2.13.0+cu130, CUDA 13.0, fa3 attention backend.

### PR DESCRIPTION
**Pull request by yuweih205**
## Motivation

Fixes #37059.

MiMo-V2.5 can deadlock on an audio request when it is served with TP8/DP2 and DP attention. The audio-tokenizer encoder attention weights are sharded over the attention-TP group, but its output projection defaults to reducing over the full TP group. When an audio request is routed to one DP rank, the other DP group never enters that collective, so the server wedges. If both DP groups contain audio, the same full-TP reduction can also mix unrelated batches.

## Changes

- Pass `use_dp_attention_reduce=is_dp_attention_enabled()` to the audio encoder's `VisionAttention`, so the output projection reduces over the same attention-TP group as its sharded weights. Non-DP-attention behavior is unchanged.
- Rebase the change onto current `main` at `ccf9fe6590ee7437005d8353c3c67d2dc4d25fcb` (candidate commit `a565eac941c89d993c64ec5d20637c79547545f8`).
- Extend the existing registered MiMo-V2 test on `base-c / 8-gpu-h200` with one audio request pinned to `routed_dp_rank=0`. The test uses `max_tokens=1` and a 120-second client timeout, checks that the request returns HTTP 200 with one choice, and verifies that `prompt_tokens_details.audio_tokens` is greater than zero so it cannot pass without exercising the audio path. Model and audio paths can be overridden for offline test environments.

`mimo_v2_asr.py` inherits the fix through `AudioEncoderMixin`.

### Existing 8×H200 validation

The same fix was previously tested with `--tp 8 --dp 2 --enable-dp-attention --enable-dp-lm-head --mm-enable-dp-encoder`. Each post-fix run covered six AV cases with two samples each: text only, audio only, video without audio, video with an audio track, video with a short audio track, and video plus a separate WAV file.

| Configuration | Before | After |
|---|---|---|
| TP8/DP2, official FP8 | The first audio request hung for 600 seconds and the NCCL watchdog killed the server | 12/12 cases passed; 0 NCCL watchdog timeouts |
| TP8/DP2, BF16 export | The first audio request reproduced the same hang | 12/12 cases passed; 0 NCCL watchdog timeouts |
| TP4/DP1, FP8 control | Not affected by the DP-attention bug | 12/12 cases passed; 0 NCCL watchdog timeouts |

Audio transcription from the fixed TP8/DP2 runs matched the TP4/DP1 reference output.

### Current-main registered regression

Candidate `a565eac941c89d993c64ec5d20637c79547545f8` (parent `ccf9fe6590ee7437005d8353c3c67d2dc4d25fcb`) passed the registered single-audio regression on 8×H200 with TP8/DP2 and DP attention. The request pinned to DP rank 0 returned HTTP 200 with one choice and a positive `prompt_tokens_details.audio_tokens` count; no NCCL collective-timeout signature appeared in the complete log. The test took 962.276 seconds including model loading, CUDA initialization, and server shutdown. The job verified the candidate archive and both changed-file SHA-256 hashes before launch.

## Checklist

- [x] Rebased onto current `main`
- [x] Added a registered multi-DP-group audio regression test
- [x] Documented existing FP8, BF16, and non-DP H200 validation
- [x] Record the current-main 8×H200 regression result
- [x] Documentation changes are not needed

---
### CI States

Latest PR Test (Base): :x: [Run #34452233269](https://github.com/sgl-project/sglang/actions/runs/34452233269)
Latest PR Test (Extra): :x: [Run #34452233147](https://github.com/sgl-project/sglang/actions/runs/34452233147)
Latest PR Test (AMD ROCm 10): :x: [Run #34452233204](https://github.com/sgl-project/sglang/actions/runs/34452233204)

### CODE REVIEW

**Comment by yuweih205** [comment]
Two additional verification rounds on the same 8×H200 machine, strengthening the original report:

**1. Second independent confirmation of the fix, different dtype & weights.** Same DP config (`--tp 8 --dp 2 --enable-dp-attention --enable-dp-lm-head --mm-enable-dp-encoder`), but serving a **bf16** fine-tuned export (`--dtype bfloat16`) instead of the official FP8: full 6-case AV suite passes **12/12**, zero NCCL watchdog timeouts, audio transcriptions match the TP4 reference verbatim. Before the fix this exact config hung on the first audio request with both FP8 and bf16 weights — so the fix is confirmed independent of quantization and checkpoint.

**2. Non-DP regression: behavior unchanged.** Fixed tree with `--tp 4` (dp=1, `is_dp_attention_enabled()` = False, i.e. the flag resolves to the pre-fix behavior): full suite **12/12**, zero NCCL timeouts. As expected from the one-line change, configurations without DP attention are untouched.

Summary across all runs of the fixed tree:

| config | dtype | result |
|---|---|---|
| tp8 dp2 + DP attention (the failing config) | fp8 | 12/12, 0 NCCL timeouts |
| tp8 dp2 + DP attention | bf16 | 12/12, 0 NCCL timeouts |
| tp4 dp1 (regression control) | fp8 | 12/12, 0 NCCL timeouts |

We have standing access to this 8×H200 + MiMo-V2.5 environment (official FP8 and bf16 exports) and are happy to run any additional configuration reviewers would like to see.

**Comment by yuweih205** [comment]
@Abatom @ShangmingCai Could a maintainer add the `run-ci` label so CI can run? The red gate here is only the missing label, not a test failure.

This fixes a deadlock we hit in production serving MiMo-V2.5 with `--tp 8 --dp 2 --enable-dp-attention`: any request containing audio wedges the schedulers until the watchdog kills the server (full analysis and repro in #37059 — the audio encoder shards attention over the attn-TP group but all-reduces over the full TP group). With this one-line fix we've verified on 8×H200: official FP8 and bf16 checkpoints, 12/12 multimodal cases (text / video / audio / mixed), no NCCL timeouts. Happy to address any review feedback.

**Comment by yuweih205** [comment]
Hi @hnyls2002 @hzh0425 — sorry for the direct ping; you two seem to handle `run-ci` labels most often lately.

Three small MiMo-V2.5 fixes are currently blocked only by the label gate (all red checks are the `pr-gate` label check itself; the 90 real test jobs are skipped):

- #37060 (this one): one-line fix for a scheduler deadlock when serving MiMo-V2.5 with `--tp 8 --dp 2 --enable-dp-attention` and any audio input — full analysis in #37059, verified on 8×H200 (fp8 + bf16, 12/12 multimodal cases, plus an independent second verification round in the comments);
- #37063: retry for the `freeze_gc` startup race under `--skip-server-warmup`;
- #37064: warning when `--dtype auto` silently falls back to fp16 for configs with no declared torch_dtype.

Could you add `run-ci` so CI can actually run? Happy to rebase or address any review feedback. Thanks!
